roles/v2/pool/src/lib/mining_pool/mod.rs

```diff
@@ -178,78 +197,99 @@ impl Pool {
+                        status_tx,
+                        Self::accept_incoming_connection_(self_.clone(), receiver, sender).await
+                    );
+                }
+                Err(_e) => {
+                    todo!()
+                }
+            }
         }
     }
 
     async fn accept_incoming_connection_(
         self_: Arc<Mutex<Pool>>,
         receiver: Receiver<EitherFrame>,
         sender: Sender<EitherFrame>,
-    ) {
-        let solution_sender = self_.safe_lock(|p| p.solution_sender.clone()).unwrap();
-
-        let channel_factory = self_.safe_lock(|s| s.channel_factory.clone()).unwrap();
+    ) -> PoolResult<()> {
+        let solution_sender = self_.safe_lock(|p| p.solution_sender.clone())?;
+        let status_tx = self_.safe_lock(|s| s.status_tx.clone())?;
+        let channel_factory = self_.safe_lock(|s| s.channel_factory.clone())?;
 
         let downstream = Downstream::new(
             receiver,
             sender,
             solution_sender,
             self_.clone(),
             channel_factory,
+            // convert Listener variant to Downstream variant
+            status_tx.listener_to_connection(),
         )
-        .await;
+        .await?;
 
-        let (_, channel_id) = downstream
-            .safe_lock(|d| (d.downstream_data.header_only, d.id))
-            .unwrap();
+        let (_, channel_id) = downstream.safe_lock(|d| (d.downstream_data.header_only, d.id))?;
 
-        self_
-            .safe_lock(|p| {
-                p.downstreams.insert(channel_id, downstream);
-            })
-            .unwrap();
+        self_.safe_lock(|p| {
+            p.downstreams.insert(channel_id, downstream);
+        })?;
+        Ok(())
     }
 
     async fn on_new_prev_hash(
         self_: Arc<Mutex<Self>>,
         rx: Receiver<SetNewPrevHash<'static>>,
         sender_message_received_signal: Sender<()>,
     ) {
+        let status_tx = self_.safe_lock(|s| s.status_tx.clone()).unwrap();
         while let Ok(new_prev_hash) = rx.recv().await {
             debug!("New prev hash received: {:?}", new_prev_hash);
-            self_
+            let res = self_
                 .safe_lock(|s| {
                     s.last_prev_hash_template_id = new_prev_hash.template_id;
                 })
-                .unwrap();
-            let x = self_
+                .map_err(|e| PoolError::PoisonLock(e.to_string()));
+            handle_result!(status_tx, res);
+
+            let x_res = self_
                 .safe_lock(|s| {
                     s.channel_factory
                         .safe_lock(|f| f.on_new_prev_hash_from_tp(&new_prev_hash))
-                        .unwrap()
+                        .map_err(|e| PoolError::PoisonLock(e.to_string()))
                 })
-                .unwrap();
+                .map_err(|e| PoolError::PoisonLock(e.to_string()));
+            let x_res = handle_result!(status_tx, x_res);
+            let x = handle_result!(status_tx, x_res);
```

Comment:
nitpick - can you change the name of this variable to something descriptive of its content - maybe job_id_res